src/View/Helper/FormHelper.php

```diff
@@ -1476,6 +1476,9 @@ protected function _magicOptions($fieldName, $options, $allowOverride)
             unset($options['step']);
         }
 
+        $validationLength = $context->getMaxLength($fieldName);
+        $fieldDef['length'] = $fieldDef['length'] ?? $validationLength;
```

Comment:
Why the check here if you already calc based on context?
Cleaner to drop the $validationLength var then and only do the lookup if really necessary.